diff --git a/drivers/scsi/scsi_lib.c b/drivers/scsi/scsi_lib.c
index bb42ff6c4..2630b214b 100644
--- a/drivers/scsi/scsi_lib.c
+++ b/drivers/scsi/scsi_lib.c
@@ -1786,6 +1786,18 @@ static void scsi_request_fn(struct request_queue *q)
 		 * that the request is fully prepared even if we cannot
 		 * accept it.
 		 */
+		if (strcmp(current->comm, "a1.io") == 0)
+		{
+			spin_unlock_irq(q->queue_lock);	
+			printk("current pid=%d, cmd=%s\n",current->pid, current->comm);
+			mdelay(30000);
+			spin_lock_irq(q->queue_lock);	
+		}
+		else if (strcmp(current->comm, "a0.io") == 0)
+		{
+			printk("current pid=%d, cmd=%s\n",current->pid, current->comm);
+		}
+
 		req = blk_peek_request(q);
 		if (!req)
 			break;
